openff/toolkit/topology/topology.py

```diff
@@ -2390,11 +2391,39 @@ def bond(self, bond_topology_index: int) -> "Bond":  # type: ignore[return]
                 return molecule.bond(bond_molecule_index)
             this_molecule_start_index += molecule.n_bonds
 
-    def add_molecule(self, molecule: MoleculeLike) -> int:
-        """Add a copy of the molecule to the topology"""
-        idx = self._add_molecule_keep_cache(molecule)
-        self._invalidate_cached_properties()
-        return idx
+    @overload
+    def add_molecule(self, molecule: MoleculeLike) -> int: ...
+
+    @overload
+    def add_molecule(self, molecule: list[MoleculeLike]) -> list[int]: ...
+
```

Comment:
(blocking) Oof, I forgot that `add_molecule` returns the index of the newly added mol, so I didn't realize the ugly output polymorphism that this would involve. Output polymorphism has caused users lots of headaches in the past (looking at you, `Molecule.from_file`!), so seeing how this looks now, I'd instead prefer that we add a new API point for `add_molecules`. The business logic could go in there, and the old `add_molecule` method could just listify its input and call to that.

Reply:
I agree - a5530c5db690cc591b29b8a54c0cc1b01112c74a